cpp/include/cudf/detail/copy_if_else.cuh

```diff
@@ -0,0 +1,197 @@
+/*
+ * Copyright (c) 2019, NVIDIA CORPORATION.
+ *
+ * Licensed under the Apache License, Version 2.0 (the "License");
+ * you may not use this file except in compliance with the License.
+ * You may obtain a copy of the License at
+ *
+ *     http://www.apache.org/licenses/LICENSE-2.0
+ *
+ * Unless required by applicable law or agreed to in writing, software
+ * distributed under the License is distributed on an "AS IS" BASIS,
+ * WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+ * See the License for the specific language governing permissions and
+ * limitations under the License.
+ */
+
+#pragma once
+
+#include <cudf/cudf.h>
+#include <cudf/column/column.hpp>
+#include <cudf/column/column_view.hpp>
+#include <cudf/column/column_device_view.cuh>
+#include <cudf/utilities/type_dispatcher.hpp>
+#include <cudf/detail/utilities/cuda.cuh>
+#include <utilities/bit_util.cuh>
+#include <utilities/cuda_utils.hpp>
+
+#include <rmm/device_scalar.hpp>
+#include <cub/cub.cuh>
+
+
+namespace cudf {
+namespace experimental {
+namespace detail {
+
+namespace {    // anonymous
+
+template <size_type block_size, typename T, typename Filter, bool has_validity>
+__global__
+void copy_if_else_kernel(  column_device_view const lhs,
+                           column_device_view const rhs,
+                           Filter filter,
+                           mutable_column_device_view out,
+                           size_type * __restrict__ const valid_count)
+{   
+   const cudf::size_type tid = threadIdx.x + blockIdx.x * blockDim.x;   
+   const int w_id = tid / warp_size;
+   // begin/end indices for the column data
+   size_type begin = 0;
+   size_type end = lhs.size();   
+   // warp indices.  since 1 warp == 32 threads == sizeof(bit_mask_t) * 8,
+   // each warp will process one (32 bit) of the validity mask via 
+   // __ballot_sync()
+   size_type w_begin = cudf::util::detail::bit_container_index<bit_mask::bit_mask_t>(begin);
+   size_type w_end = cudf::util::detail::bit_container_index<bit_mask::bit_mask_t>(end);
+
+   // lane id within the current warp
+   constexpr size_type w_leader_lane{0};
+   const int w_lane_id = threadIdx.x % warp_size;
+
+   size_type w_valid_count{0};
+  
+   // current warp.
+   size_type w_cur = w_begin + w_id; 
+   auto w_active_mask = __ballot_sync(0xFFFF'FFFF, w_cur <= w_end);
+   while(w_cur <= w_end){
+      // absolute element index
+      size_type index = (w_cur * warp_size) + w_lane_id;
+      bool in_range = (index >= begin && index < end);
+
+      bool valid = true;
+      if(has_validity){
+         valid = in_range && filter(index) ? lhs.is_valid(index) : rhs.is_valid(index);
+      }
+
+      // do the copy if-else, but only if this element is valid in the column to be copied 
+      if(in_range && valid){ 
```

Comment:
I'd avoid the divergence and just do the copy anyways even if the element is null.